src/middlewared/middlewared/main.py

```diff
@@ -1368,6 +1370,57 @@ def log_threads_stacks(self):
         for thread_id, stack in get_threads_stacks().items():
             self.logger.debug('Thread %d stack:\n%s', thread_id, ''.join(stack))
 
+    def _tracemalloc_start(self):
+        """
+        Run an endless loop grabbing snapshots of allocated memory using
+        the python's builtin "tracemalloc" module.
+        """
+        # set the thread name
+        osc.set_thread_name('tracemalloc_monitor')
+
+        # initalize tracemalloc
+        tracemalloc.start()
+
+        # change this to limit the number of lines
+        # showing the highest memory consumer
+        limit = 5
+
+        # how often to sleep in seconds before
+        # taking another memory snapshot
+        interval = 5
+
+        # filters for the snapshots so we can
+        # ignore modules that we don't care about
+        filters = (
+            tracemalloc.Filter(False, '<frozen importlib._bootstrap>'),
+            tracemalloc.Filter(False, '<frozen importlib._bootstrap_external>'),
+            tracemalloc.Filter(False, '<unknown>'),
+            tracemalloc.Filter(False, '*tracemalloc.py'),
+        )
+
+        # start the loop
+        prev = None
+        while True:
+            if prev is None:
+                prev = tracemalloc.take_snapshot()
+                prev = prev.filter_traces(filters)
+            else:
+                curr = tracemalloc.take_snapshot()
+                curr = curr.filter_traces(filters)
+                diff = curr.compare_to(prev, 'lineno')
+
+                prev = curr
+                curr = None
+                for idx, stat in enumerate(diff[:limit], 1):
+                    self.logger.debug(f'#{idx}: {stat}')
```

Comment:
This can be potentially prone to race conditions with other loggers (tracemalloc logs interleaved with something else), maybe we should log entire record at once?